### PR TITLE
refactor dialogue state tracking for modelling/dataset interoperability

### PR DESCRIPTION
**Pull request by Zhilin123**
Signed-off-by: Zhilin Wang <zhilinw@nvidia.com>

Design based on attached diagram
[UML proposed dialogue1.pdf](https://github.com/NVIDIA/NeMo/files/7946212/UML.proposed.dialogue1.pdf)

Currently supports SGD-QA (Bert-based) and Huggingface GPT2 models, and planning to support other models.

### CODE REVIEW

**Comment by Zhilin123** [comment]
Integrate testing done by comparing output of 1. modified Dataset/DataProcessor with original SGDQA model and 2. original Dataset/DataProcessor with original SGDQA model. Both give identical output

`CUDA_VISIBLE_DEVICES=0 python ~/NeMo/examples/nlp/dialogue_state_tracking_generative/sgd_gen.py do_training=False trainer.gpus=1 model.dataset.data_dir=/home/zhilinw/dstc8-schema-guided-dialogue/ model.dataset.dialogues_example_dir=/home/zhilinw/dstc8-schema-guided-dialogue/ model.dataset.use_cache=False`

and 

`CUDA_VISIBLE_DEVICES=0 python ~/NeMo/examples/nlp/dialogue_state_tracking/sgd_qa.py do_training=False trainer.gpus=1 model.dataset.data_dir=/home/zhilinw/dstc8-schema-guided-dialogue/ model.dataset.dialogues_example_dir=/home/zhilinw/dstc8-schema-guided-dialogue/ model.dataset.use_cache=False`

`[NeMo I 2022-01-25 16:32:03 evaluate:284] Dialog metrics for #SEEN_SERVICES  : [('active_intent_accuracy', 38.36), ('average_cat_accuracy', 13.69), ('average_cat_status_accuracy', 100.0), ('average_cat_value_accuracy', 13.69), ('average_goal_accuracy', 25.8), ('average_noncat_accuracy', 30.83), ('average_noncat_status_accuracy', 98.83), ('average_noncat_value_accuracy', 31.19), ('joint_cat_accuracy', 4.19), ('joint_cat_status_accuracy', 47.69), ('joint_cat_value_accuracy', 9.39), ('joint_goal_accuracy', 0.25), ('joint_noncat_accuracy', 4.59), ('joint_noncat_status_accuracy', 22.48), ('joint_noncat_value_accuracy', 13.44), ('requested_slots_f1', 2.7), ('requested_slots_precision', 3.05), ('requested_slots_recall', 90.75)]
[NeMo I 2022-01-25 16:32:03 evaluate:286] Dialog metrics for #UNSEEN_SERVICES: [('active_intent_accuracy', 33.82), ('average_cat_accuracy', 18.37), ('average_cat_status_accuracy', 100.0), ('average_cat_value_accuracy', 18.37), ('average_goal_accuracy', 25.22), ('average_noncat_accuracy', 27.62), ('average_noncat_status_accuracy', 97.85), ('average_noncat_value_accuracy', 28.22), ('joint_cat_accuracy', 1.18), ('joint_cat_status_accuracy', 25.6), ('joint_cat_value_accuracy', 11.14), ('joint_goal_accuracy', 0.51), ('joint_noncat_accuracy', 2.22), ('joint_noncat_status_accuracy', 42.34), ('joint_noncat_value_accuracy', 9.29), ('requested_slots_f1', 4.21), ('requested_slots_precision', 4.25), ('requested_slots_recall', 91.42)]
[NeMo I 2022-01-25 16:32:03 evaluate:288] Dialog metrics for #ALL_SERVICES   : [('active_intent_accuracy', 34.85), ('average_cat_accuracy', 17.2), ('average_cat_status_accuracy', 100.0), ('average_cat_value_accuracy', 17.2), ('average_goal_accuracy', 25.35), ('average_noncat_accuracy', 28.36), ('average_noncat_status_accuracy', 98.08), ('average_noncat_value_accuracy', 28.91), ('joint_cat_accuracy', 1.92), ('joint_cat_status_accuracy', 31.04), ('joint_cat_value_accuracy', 10.7), ('joint_goal_accuracy', 0.45), ('joint_noncat_accuracy', 2.76), ('joint_noncat_status_accuracy', 37.84), ('joint_noncat_value_accuracy', 10.25), ('requested_slots_f1', 3.87), ('requested_slots_precision', 3.98), ('requested_slots_recall', 91.27)]
`

File: nemo/collections/nlp/modules/common/gpt_module.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            return
+
+        pretrained_dict = torch.load(restore_path)
+
+        # backward compatibility with NeMo0.11
+        if "state_dict" in pretrained_dict.keys():
+            pretrained_dict = pretrained_dict["state_dict"]
+
+        # remove prefix from pretrained dict
+        m = re.match("^bert.*?\.", list(pretrained_dict.keys())[0])
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/gpt_module.py:66]: is this a bug? shouldn't this be gpt instead of bert

Comment (Zhilin123) [reply nemo/collections/nlp/modules/common/gpt_module.py:68]: fixed

**Comment by okuchaiev** [comment]
ping @cparisien and @yzhang123

**Review by cparisien — APPROVED** [review]
This looks good to me from a design perspective. We should still get a review from someone who is a regular nemo contributor, since I'm not.

File: nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+            dialogues_processor: Data generator for SGD dialogues
+        """
+        self.dataset_split = dataset_split
+        self.features = []
+        self.tokenizer = tokenizer
+        self.schemas = schemas
+        self.schema_config = schema_config
+        self.raw_features = dialogues_processor.get_dialog_examples(dataset_split)
+
+        for idx in range(len(self.raw_features)):
```
Comment (yzhang123) [inline nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py:64]: have you tested how long this takes on SGD? since previously we used to cache the tokenized dataset, now we need to compute it from examples dict

File: nemo/collections/nlp/models/dialogue_state_tracking_hybrid/sgdqa_model.py [hunk 3]
```diff
@@ -0,0 +1,623 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+# Copyright 2019 The Google Research Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
```
Comment (yzhang123) [inline nemo/collections/nlp/models/dialogue_state_tracking_hybrid/sgdqa_model.py:9]: Could you make sure this ckpt compatible with the new code? https://ngc.nvidia.com/models/nvstaging:nemo:sgdqa_bertbasecased

**Review by yzhang123 — COMMENTED** [review]
thanks for the nice PR! Looks good to me overall. Left some comments, 
Could you update the description with all the file/class respondences?

SGDDataProcessor -> DialogueSGDDataProcessor
etc

Please  
1. check if the old model is compatible 
2. cache and load from cache the tokenized dataset (like before)
3. verify training of SGD model still reaches previous accuracy on 8-GPUs

afterwards we can replace  old code with this new folder to avoid redundancy

Thanks!

**Review by yzhang123 — CHANGES_REQUESTED** [review]
left some comments

File: Jenkinsfile [hunk 4]
```diff
@@ -663,23 +663,68 @@ pipeline {
...
+            trainer.gpus=[0] \
+            model.dataset.use_cache=false \
+            model.tokenizer.special_tokens={pad_token:"<|endoftext|>"}\
+            model.language_model.pretrained_model_name=gpt2 \
+            trainer.accelerator=ddp \
+            exp_manager=null  && \
+            rm -rf sgd_gen_outputs'
+          }
+        }
+        stage('SGD-GEN Backward compatible with bert-base-cased') {
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply Jenkinsfile:708]: Updated line 716 from `model.language_model.pretrained_model_name=gpt2` to `model.language_model.pretrained_model_name=bert-base-cased`

File: nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py (hunk 2 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py:64]: It takes around 7 minutes to process this. Since this takes some time, I have now implemented the caching function as part of the dataset class.

File: nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/prediction_utils.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+Prediction and evaluation-related utility functions.
+This file contains code artifacts adapted from the original implementation:
+https://github.com/google-research/google-research/blob/master/schema_guided_dst/baseline/pred_utils.py
+"""
+
+import json
+import os
+from collections import OrderedDict, defaultdict
+from typing import Dict, List, Optional
+
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/prediction_utils.py:26]: updated!

File: nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/data_processor.py [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+        """
+        Obtain the list of all dialogues from specified json files.
+        Args:
+            dialog_json_filepaths: list of json files
+        Returns:
+            dialogs: the list of all dialogues
+        """
+        dialogs = []
+        for dialog_json_filepath in sorted(dialog_json_filepaths):
+            with open(dialog_json_filepath, 'r') as f:
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/data_processor.py:507]: updated!

File: nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/input_example.py [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+
+    def __repr__(self):
+        return self.data
+
+    def __str__(self):
+        return self.data
+
+
+class DialogueSGDInputExample(DialogueInputExample):
+    pass
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/input_example.py:62]: This class contains the description for how an instance of `DialogueSGDInputExample` looks like in Lines 63 to 110 (right below). As it doesn't override any methods from its parent class, I simply wrote pass.

File: nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+                                slot_inv_alignments,
+                                system_user_tokens,
+                                system_user_inv_alignments,
+                                slot_description,
+                                system_user_utterance,
+                            )
+                            task_example.add_categorical_slots(state_update)
+                            assert task_example.categorical_slot_status == old_example.categorical_slot_status
+                            examples.append(task_example)
+
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py:298]: added this!

File: nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+                        examples.append(task_example)
+
+                    if self.dataset_split != 'train' or task_example.noncategorical_slot_status == 1:
+                        task_example = task_example.make_copy_of_non_categorical_features()
+                        task_example.task_mask[5] = 1
+                        # assert task_example.task_mask == [0, 0, 0, 0, 0, 1]
+                        task_example.example_id = base_example.example_id + f"-5-{slot_id}-0"
+                        task_example.example_id_num = base_example.example_id_num + [5, slot_id, 0]
+                        examples.append(task_example)
+
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py:346]: added this!

File: nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+                    user_span_boundaries = self._find_subword_indices(
+                        state_update,
+                        user_utterance,
+                        ex["label_positions"]["slots"],  # user_frame["slots"],
+                        user_alignments,
+                        user_tokens,
+                        2 + len(slot_tokens) + len(system_tokens),
+                    )
+
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue_state_tracking_generative/sgd/dialogue_bert_dataset.py:330]: added!

File: nemo/collections/nlp/models/dialogue_state_tracking_hybrid/sgdqa_model.py (hunk 3 above)
Comment (Zhilin123) [reply nemo/collections/nlp/models/dialogue_state_tracking_hybrid/sgdqa_model.py:9]: This shouldn't be a problem as I only changed the Dataset class and .nemo doesn't save the dataset information if I understand it correctly. Nonetheless, if you would like me to double check this, can you send a link on how I can check for this (not too sure how I can load the model from ngc) ?

**Comment by Zhilin123** [comment]
> 3. verify training of SGD model still reaches previous accuracy on 8-GPUs

Ran both the SGDQA and SGDGEN models on 2-gpu setup 

SGD-GEN
Screen Shot 2022-02-08 at 2 59 50 PM

SGD-QA
Screen Shot 2022-02-08 at 8 21 55 PM

Details of weights and biases see attached pdf

[Comparing SGDQA and SGD-GEN _ SGDGEN – Weights & Biases.pdf](https://github.com/NVIDIA/NeMo/files/8029487/Comparing.SGDQA.and.SGD-GEN._.SGDGEN.Weights.Biases.pdf)

**Comment by Zhilin123** [comment]
> Could you update the description with all the file/class respondences?
> 
> SGDDataProcessor -> DialogueSGDDataProcessor etc

@yzhang123 Can you be more specific about what this means? The class `DialogueSGDDataProcessor ` has already been created for instance.

**Review by yzhang123 — DISMISSED** [review]

**Review by yzhang123 — APPROVED** [review]
